# List of some alternative tools (additions to this list are welcome): 

This is in no way an exhaustive list of all the available resources. Feel free to suggest other resources that can be added to this list as it will help users to select the tools that are most suitable for their needs. 

### Standalone programs and web browsers for GWAS visualization: 

[WGAViewer](http://igm.cumc.columbia.edu/WGAViewer/), Ge et al, 2008  
[LocusZoom](http://locuszoom.sph.umich.edu/locuszoom/)* Pruim et al, 2010  
[GWAS Central](http://www.gwascentral.org/) * Beck et al, 2014  
[GWAPP](http://gwapp.gmi.oeaw.ac.at/index.html#!homePage)* Seren et al, 2012  
[GWASrap](http://jjwanglab.org/gwasrap)* Li et al, 2012  
[JBrowse](http://jbrowse.org/)* Skinner et al, 2009  
[ZBrowse](http://www.baxterlab.org/#!untitled/cqi0)* Ziegler et al, 2015  
*comparison for these tools presented in Zeigler et al, 2015

### Packages (most are written in R) and other code:
1. [qqman] (https://github.com/stephenturner/qqman) [Turner, 2014](http://biorxiv.org/content/early/2014/05/14/005165)  
2. Code samples for generating [QQ] (http://genome.sph.umich.edu/wiki/Code_Sample:_Generating_QQ_Plots_in_R#Under_the_Hood:_Confidence_Intervals) and [Manhattan] (http://genome.sph.umich.edu/wiki/Code_Sample:_Generating_Manhattan_Plots_in_R) in R from Abecasis Lab (author: Matthew Flickinger)  
3. ggbio: visualization toolkits for genomic data, Manhattan for GWAS, [plotGrandLinear](http://rpackages.ianhowson.com/bioc/ggbio/man/plotGrandLinear.html) (authors: Tengfei Yin, Dianne Cook, Michael Lawrence, 2016). [Manual](http://bioconductor.org/packages/3.3/bioc/vignettes/ggbio/inst/doc/ggbio.pdf)  
4. [mhtplot](http://www.inside-r.org/packages/cran/gap/docs/mhtplot) (author: Jing Hua Zhao)
5. GWAS Tools, Functions: manhattanPlot and qqPlot. [Manual](https://bioconductor.org/packages/3.3/bioc/manuals/GWASTools/man/GWASTools.pdf)
6. Plotting Genome-Wide Association Results, [Q-Q Plot and regional association plot](https://www.broadinstitute.org/diabetes/scandinavs/figures.html)
7. R Graph Gallery, RG#27: [Manhattan plot]( http://rgraphgallery.blogspot.com/2013/04/manhattan-plot.html)

#### References
Beck T, Hastings RK, Gollapudi S, Free RC, Brookes AJ. 2014. GWAS Central: a comprehensive resource for the comparison and interrogation of genome-wide association studies. European Journal of Human Genetics 22:949–952 DOI 10.1038/ejhg.2013.274.  

Cuellar-Partida G, Renteria ME,MacGregor S. 2015. LocusTrack: integrated visualization of GWAS results and genomic annotation. Source Code for Biology and Medicine 10:1–4. DOI 10.1186/s13029-015-0032-8.  

Ge D, Zhang K, et al. 2008. WGAViewer: software for genomic annotation of whole genome association studies. Genome Research 18 (4): 640–643.  

Li MJ, Sham PC,Wang J. 2012. Genetic variant representation, annotation and prioritization in the post-GWAS era. Cell Research 22:1505–1508 DOI 10.1038/cr.2012.106.  

Pruim RJ, et al. 2010. LocusZoom: Regional visualization of genome-wide association scan results. Bioinformatics, 26(18): 2336.2337.  

Seren ¨ U, Vilhj´almsson BJ, Horton MW,Meng D, Forai P, Huang YS, Long Q, Segura V, Nordborg M. 2012. GWAPP: a web application for genome-wide association mapping in Arabidopsis. The Plant Cell Online 24:4793–4805 DOI 10.1105/tpc.112.108068.  

Skinner ME, Uzilov AV, Stein LD,Mungall CJ, Holmes IH. 2009. JBrowse: a next-generation genome browser. Genome Research 19:1630–1638 DOI 10.1101/gr.094607.109.  

Turner, SD. 2014 qqman: an R package for visualizing GWAS results using Q-Q and Manhattan plots. bioRxiv, doi: 10.1101/005165.

Zeigler GR, Hartsock RH, and Baxter I. 2015. Zbrowse: an interactive GWAS results browser. PeerJ Comput. Sci. DOI 10.7717/peerj-cs.3



